File: php/1448-count-good-nodes-in-binary-tree.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer
     */
    function goodNodes($root) {
        if ($root === null) return 0;

        $count = 0;
        $stack = [[$root->val, $root]];
        while (count($stack) > 0) {
            $current = array_pop($stack);
            $max = $current[0];
            $node = $current[1];
            if ($node->val >= $max) {
                $max = $node->val;
                $count = $count + 1;
            }
            if ($node->right !== null) array_push($stack, [$max, $node->right]);
            if ($node->left !== null) array_push($stack, [$max, $node->left]); 
        }
        return $count;
    }
}

File: golang/1448-count-good-nodes-in-binary-tree.go
package golang

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type MaxNode struct {
	Tree     *TreeNode
	MaxSoFar int
}

func goodNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}

	count := 0
	stack := []MaxNode{{Tree: root, MaxSoFar: root.Val}}
	for len(stack) > 0 {
		current := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		maxSoFar := current.MaxSoFar
		if current.Tree.Val >= current.MaxSoFar {
			count++
			maxSoFar = current.Tree.Val
		}
		if current.Tree.Right != nil {
			stack = append(stack, MaxNode{Tree: current.Tree.Right, MaxSoFar: maxSoFar})
		}
		if current.Tree.Left != nil {
			stack = append(stack, MaxNode{Tree: current.Tree.Left, MaxSoFar: maxSoFar})
		}
	}
	return count
}
